contracts/token/ERC721/extensions/ERC721Wrapper.sol

```diff
@@ -0,0 +1,55 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+import "../ERC721.sol";
+import "../utils/ERC721Holder.sol";
+
+/**
+ * @dev Extension of the ERC721 token contract to support token wrapping.
+ *
+ * Users can deposit and withdraw an "underlying token" and receive a "wrapped token" with a matching tokenId. This is useful
+ * in conjunction with other modules. For example, combining this wrapping mechanism with {ERC721Votes} will allow the
+ * wrapping of an existing "basic" ERC721 into a governance token.
+ *
+ * _Available since v4.9.0_
+ */
+abstract contract ERC721Wrapper is ERC721, ERC721Holder {
+    IERC721 public immutable underlying;
+
+    constructor(IERC721 underlyingToken) {
+        underlying = underlyingToken;
+    }
+
+    /**
+     * @dev Allow a user to deposit underlying tokens and mint the corresponding tokenIds.
+     */
+    function depositFor(address account, uint256[] memory tokenIds) public virtual returns (bool) {
+        for (uint256 i = 0; i < tokenIds.length; ++i) {
+            underlying.safeTransferFrom(_msgSender(), address(this), tokenIds[i]);
+            _safeMint(account, tokenIds[i]);
+        }
+        return true;
+    }
+
+    /**
+     * @dev Allow a user to burn wrapped tokens and withdraw the corresponding tokenIds of the underlying tokens.
+     */
+    function withdrawTo(address account, uint256[] memory tokenIds) public virtual returns (bool) {
```

Comment:
WOW, this function doesn't have any sort of access restriction!

We need to check that `_msgSender` is `isApprovedOrOwner`

Reply:
Same for [ERC20Wrapper](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/extensions/ERC20Wrapper.sol#L48). The check is assumed to be performed in `safeTransferFrom`

Reply:
No, 

`underlying.safeTransferFrom` will check that the caller (this contract) is approved of owner of the underlying token ... which is the case because the underlying token is held by the wrapper.

What you need to check is that the caller has authority over the the wrapped token.

In ERC20Wrapper, this is done by burning from the _msgSender() balance. You can only burn your token to unwrap them. Here anyone can burn any token, and unwrap them to any token

Reply:
Just add a test where an account tries to unwrap a token it doesn't own

Reply:
Ahh got it, thanks for flagging. I assumed _burn was restricted for some reason.
Added also tests accordingly.